icu4c/source/common/uloc.cpp

```diff
@@ -1465,6 +1465,25 @@ uloc_openKeywords(const char* localeID,
 static const char i_default[] = {'i', '-', 'd', 'e', 'f', 'a', 'u', 'l', 't'};
 #define I_DEFAULT_LENGTH UPRV_LENGTHOF(i_default)
 
+static int32_t _preprocessLocaleID(const char* inLocaleID, char* outBuffer, int32_t bufferCapacity, UErrorCode* err) {
+    // this function is to allow permissive parsing in locale IDs in general cases, where the locale ID can be in BCP
+    // 47 format, but with underbars as field delimiters
+    int32_t neededLength = static_cast<int32_t>(uprv_strlen(inLocaleID));
+    if (neededLength > bufferCapacity) {
+        *err = U_BUFFER_OVERFLOW_ERROR;
+        return neededLength;
+    }
+    
+    uprv_strcpy(outBuffer, inLocaleID);
+    if (uprv_strchr(inLocaleID, '_') != nullptr && ((inLocaleID[0] != 'x' && inLocaleID[0] != 'X') || (inLocaleID[0] != '\0' && inLocaleID[1] != '-'))) {
+        // convert all underbars to hyphens, unless the string starts with "x-" and contains an underbar
+        for (char* p = outBuffer; *p != '\0'; ++p) {
+            *p = (*p == '_') ? '-' : *p;
```

Comment:
Don't write to `*p` if you don't need to modify it.

Reply:
Why?  I've changed this, but I don't understand why writing to *p hurts anything.  I did it this way for brevity.